llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -1229,14 +1360,18 @@ void RISCVInsertVSETVLI::transferBefore(VSETVLIInfo &Info,
 void RISCVInsertVSETVLI::transferAfter(VSETVLIInfo &Info,
                                        const MachineInstr &MI) const {
   if (isVectorConfigInstr(MI)) {
-    Info = getInfoForVSETVLI(MI, *MRI);
+    Info = getInfoForVSETVLI(MI, *MRI, LIS);
     return;
   }
 
   if (RISCV::isFaultFirstLoad(MI)) {
     // Update AVL to vl-output of the fault first load.
-    Info.setAVLRegDef(MRI->getVRegDef(MI.getOperand(1).getReg()),
-                      MI.getOperand(1).getReg());
+    if (MI.getOperand(1).getReg() == RISCV::X0)
+      Info.setAVLVLMAX();
```

Comment:
This is the output VL, not the AVL. If it's set to X0 then I think we lose the information about what VL might be after the vleff. That's part of the reason why it would be easier to defer RISCVDeadRegisterDefinitions till after vsetvli insertion

Reply:
So it could be remain setAVLRegDef after https://github.com/llvm/llvm-project/pull/90636 landed?

```
Info.setAVLRegDef(
    getReachingDefMI(MI.getOperand(1).getReg(), &MI, MRI, LIS),
    MI.getOperand(1).getReg());
```

Reply:
Yeah, and it would also address those regressions we're seeing where we have `vsetvli a0, a0` instead of `vsetvli x0, a0`.